### PR TITLE
[ML] Create a Docker build image using the latest PyTorch stable branch.

### PR DESCRIPTION
**Pull request by edsavage**
On Linux x86_64, provide the ability to create Docker build images that use the code pulled from the latest PyTorch stable release branch (currently 2.3.1)

Upon successful build and push of such a Docker image, trigger a build of the `ml-cpp` code in it.

Once the `ml-cpp` build has succeeded it in turn should trigger a pipeline in the QAF repo that runs a set of tests exercising the `pytorch_inference` executable.

To make this possible a new pipeline - ml-cpp-pytorch-build - is required. This is defined in the `catalog-info.yaml` file and won't be created until `catalog-info.yaml` is merged to main and backstage magic does its thing.

Some tweaks to our existing `Buildkite` framework have been made in order that existing code can be better be re-used, so that e.g. just a `linux x86_64` build step can be dynamically created.

TBD: The name of the QA PyTorch testing pipeline is required.

### CODE REVIEW

**Comment by edsavage** [comment]
To manually test the changes in this PR
  * Go to https://buildkite.com/elastic/ml-cpp-pr-builds and click on the `New Build` button.
  * Add a descriptive message in the message field
  * Add the ID of the latest commit in the commit field
  * Add the name of this PR branch - `pytorch_latest_docker_build` - in the branch field
  * Click on `Options and in the Environment variables box add:
  ```
    DOCKER_IMAGE="docker.elastic.co/ml-dev/ml-linux-dependency-build:pytorch_231"
    GITHUB_PR_COMMENT_VAR_ACTION="run_pytorch_tests"
    GITHUB_PR_COMMENT_VAR_ARCH="x86_64"
    GITHUB_PR_COMMENT_VAR_PLATFORM="linux"
    GITHUB_PR_TRIGGER_COMMENT=""
```
 * Click on `Create Build`

**Review by valeriy42 — COMMENTED** [review]
Great work on untangling the dependencies between buildkite and building docker images. I believe this work will help us identify issues much earlier.

However, I am confused about the workflow as it is intended right now. The way I understand it: 
1. PyTorch releases a new version every 4-6 weeks at the moment. Hence, the latest stable will be the same for this amount of time, so  every night we will re-build the same image and the tests will lead to the same results.
2. When PyTorch will produce a new stable branch, we will need monitor it and update the branch referenced in `dev-tools/docker/pytorch_linux_image/Dockerfile` and the image tags.

The PyTorch releases are branched of `viable/strict`. So, the way I understand the intend of this PR, we should rebuild the Docker image nightly from this PyTorch branch to identify the new issues, before we will start depending on a new PyTorch release, shouldn't we?

**Comment by edsavage** [comment]
buildkite build this

**Comment by edsavage** [comment]
> The PyTorch releases are branched of `viable/strict`. So, the way I understand the intend of this PR, we should rebuild the Docker image nightly from this PyTorch branch to identify the new issues, before we will start depending on a new PyTorch release, shouldn't we?

Thanks @valeriy42 ! Yes, the I believe the `viable/strict` branch is the most appropriate one for us to be building and testing against. Originally we had been targeting `main`, but that is too volatile. Our friends in QA suggested the latest release branch - `2.3.1`, but as you've pointed out, that is pretty much static. So I think `viable/strict` is our "Goldilocks" branch - just right 🤞

File: dev-tools/docker/README.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+the machine learning C++ code for Linux, e.g. PyTorch or Boost, in order
+to test the latest versions of them.
+
+This image is not intended to be built regularly. It is based on the latest
+`ml-linux-build` image, but with extra dependencies added in order to build
+PyTorch.  Therefore, When changing any of the tools or 3rd party components
+required to build the machine learning C++ code dependencies increment the
+version, change the Dockerfile and build a new image to be used for subsequent
+builds on this branch. Then update the version to be used for builds in docker
+files that refer to it.
```
Comment (valeriy42) [inline dev-tools/docker/README.md:36]: nit: it would be easier to parse if you would format it as a numbered list of things to do.

**Review by valeriy42 — APPROVED** [review]
LGTM. Just a minor comment.

**Pull request merged**